### DISCUSSION TITLE
How to gain insights into what poetry is working on ?

### DISCUSSION THREAD

**Discussion by dd-ssc** [python-poetry/poetry#8237] [Q&A]
First up: I am unsure where to post this; this section seems most appropriate, please let me know if there is a better place elsewhere.

### My problem

In my current (comparatively small) project, a simple `poetry install` or `poetry update` takes about 30 minutes. 
I often have to halt work and wait for poetry to finish what it is doing.

I have read [Why is the dependency resolution process slow?](https://python-poetry.org/docs/faq/#why-is-the-dependency-resolution-process-slow), several GitHub issues and other online resources - and I have a few suspicions myself about what could be the culprit, but those at the moment can only be speculative at best:
In order to improve anything, it is required to be able to properly measure the aspect to improve.

### My question

Does anyone have some advice on how to measure the time poetry is spending on its various internal steps ?

I tried adding `-vvv`, but somehow parsing that additional log output does not seem like a smart approach to me...

**Comment by radoering** [comment]
I don't have a general approach, just a few tips:

- `poetry install` might only be slow if there is no lockfile. Most of the times, locking is slow. If that's the case debug `poetry lock`, not `poetry install` or `poetry update`.
- If you compare times, always compare the second `poetry lock` (warm cache). (Unless only the first locking is slow. If you want to measure times with a cold cache you have to remove the cache to get reproducible results.)
- Try to reduce your dependencies as much as possible to identify a minimal combination of dependencies that are the culprit.
- Sometimes parsing the log output with `-vvv` gives some clue what's happening (e.g. if many versions of the same dependency are tried before finding a solution).
- Some sort of profiling can be helpful, e.g. just add some time measurements and prints at several places in the code or instrument the [lock command](https://github.com/python-poetry/poetry/blob/5571d815797da551abcbff084e09fffc7a9665a4/src/poetry/console/commands/lock.py#L55) with cProfile depending on what you want to measure (locking with cProfile will take much longer than without).

**Comment by dimbleby** [comment]
also these days the answer is always boto3 vs urllib3!  per #5896, #7858, etc
